components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/JDBCPermissionBasedInternalScopeValidator.java

```diff
@@ -326,6 +341,50 @@ private String[] getAllowedUIResourcesForNotAssociatedFederatedUser(Authenticate
         return allowedUIResourcesListForUser.toArray(new String[0]);
     }
 
+    /**
+     * Retrieve list of permissions using roles of federated user.
+     *
+     * @param authenticatedUser    FederatedAuthenticatedUser
+     * @param authorizationManager AuthorizationManager
+     * @return List of permissions
+     * @throws UserStoreException      UserStoreException
+     * @throws IdentityOAuth2Exception IdentityOAuth2Exception
+     */
+    private String[] getAllowedPermissionsUsingRoleForNonAssociatedFederatedUsers(
+            AuthenticatedUser authenticatedUser, AuthorizationManager authorizationManager)
+            throws UserStoreException, IdentityOAuth2Exception {
+
+        List<String> allowedUIResourcesListForUser = new ArrayList<>();
+        List<String> userRolesList = getRolesFromFederatedUserAttributes(authenticatedUser.getUserAttributes());
+
+        for (String  role: userRolesList) {
+            String modifiedRole = role;
+
+            // Add internal prefix before normal roles.
+            if (!modifiedRole.contains(CarbonConstants.DOMAIN_SEPARATOR)) {
+                modifiedRole = UserCoreConstants.INTERNAL_DOMAIN + CarbonConstants.DOMAIN_SEPARATOR + modifiedRole;
+            }
+
+            // Continue if it is not internal role.
+            if (!modifiedRole.startsWith(UserCoreConstants.INTERNAL_DOMAIN + CarbonConstants.DOMAIN_SEPARATOR)) {
+                continue;
+            }
+
+            // Loop through each internal local role and get permissions.
+            for (String allowedUIResource :
+                    authorizationManager.getAllowedUIResourcesForRole(modifiedRole, ROOT)) {
+                if (!allowedUIResourcesListForUser.contains(allowedUIResource)) {
+                    allowedUIResourcesListForUser.add(allowedUIResource);
```

Comment:
Can't we use a **Set** for `allowedUIResourcesListForUser`? If so we can get rid of this `contains` check.

Reply:
yes we can use